README.md
todoist-templates
================
[![pypi](https://img.shields.io/pypi/v/todoist_templates.svg)](https://pypi.python.org/pypi/todoist_templates)

Create Todoist projects from file.

Supported file types:
 * yml, yaml

How to use?
-----------
* Install `pip install --user todoist_templates`
* Set environment variable with your TodoistApi `export TODOIST_API=123123123`
* Create project template based on `example_project.yml` file
* Run `todoist_templates path/to/project_template.yml`

Supported properties
--------
All supported properties based on this api: https://developer.todoist.com/sync/v8/

* `name` - Project name
* `color` - Project color [integer] from 30 to 49 - [More explanation](https://developer.todoist.com/sync/v8/#colors)
* `tasks` - Project tasks
    * `name` - Name of task
    * `priority` - Priority of task [integer] from 1 to 4
    * `tasks` - Subtasks
* `sections` - Project sections
    * `tasks` - Tasks in section, properties are the same like in base project `tasks`

Todo
----
* Add tests
* Add JSON support


CHANGELOG.md
Changelog
=========

Unreleased
----------

0.0.1 (2020-03-21)
------------------

* First release on PyPI.


todoist_templates/cli.py
import argparse
import sys

from .api import get_api
from .errors import (ApiError, ParserError, ValidationError)
from .generator import generate_project
from .parsers import get_project
from .validator import validate_project


def _generate(path, project_name=None):
    try:
        api = get_api()
    except ApiError as error:
        print("Error:", error)
        sys.exit(1)

    try:
        project = get_project(path)
    except ParserError as error:
        print("Error:", error)
        sys.exit(1)

    try:
        validate_project(project)
    except ValidationError as error:
        print("Error:", error)
        sys.exit(1)

    if project_name:
        print('Using override project name: "{}" instead of: "{}"'.format(project_name, project['name']))
        project['name'] = project_name

    generate_project(api, project)


def main():
    parser = argparse.ArgumentParser(description='Generate Todoist project from template.')
    parser.add_argument('path', nargs='?', type=str, help='Path to template file')
    parser.add_argument('--name', type=str, help='Project name override', required=False)
    args = parser.parse_args()
    if not args.path:
        parser.print_help()
        sys.exit(1)

    _generate(args.path, project_name=args.name)


todoist_templates/generator.py
class Generator(object):
    def __init__(self, api, project):
        self.api = api
        self.project = project

    def _generate_task(self, task_template, project_id, section_id=None, parent_id=None):
        name = task_template.get('name')
        priority = task_template.get('priority')
        tasks = task_template.get('tasks', [])

        task = self.api.items.add(
            content=name,
            project_id=project_id,
            section_id=section_id,
            parent_id=parent_id,
            priority=priority
        )
        [self._generate_task(
            task_item,
            project_id=project_id,
            section_id=section_id,
            parent_id=task['id']
        ) for task_item in tasks]

    def _generate_section(self, section_template, project_id):
        name = section_template.get('name')
        tasks = section_template.get('tasks', [])

        section = self.api.sections.add(
            name=name,
            project_id=project_id
        )
        [self._generate_task(
            task_item,
            project_id=project_id,
            section_id=section['id']
        ) for task_item in tasks]

    def _generate_project(self):
        name = self.project.get('name')
        color = self.project.get('color')
        tasks = self.project.get('tasks', [])
        sections = self.project.get('sections', [])

        project = self.api.projects.add(name=name, color=color)
        [self._generate_task(task_item, project_id=project['id']) for task_item in tasks]
        [self._generate_section(section_item, project_id=project['id']) for section_item in sections]

    def generate(self, dry_run=False):
        self._generate_project()

        if not dry_run:
            self.api.commit()
        print('Project "{}" was generated'.format(self.project.get('name')))


def generate_project(api, project):
    Generator(api, project).generate()


todoist_templates/validator.py
from .errors import ValidationError


class Validator(object):
    def __init__(self, project_template):
        self.project = project_template

    def _validate_task(self, task):
        name = task.get('name')
        priority = task.get('priority')
        tasks = task.get('tasks', [])
        if not name:
            raise ValidationError('There is no "name" in one of task in your project template')
        if priority and int(priority) not in range(1, 5):
            raise ValidationError('Task with name: "{}" should have priority between: 1-4'.format(name))
        if tasks:
            [self._validate_task(task) for task in tasks]

    def _validate_section(self, section):
        name = section.get('name')
        tasks = section.get('tasks', [])
        if not name:
            raise ValidationError('There is no "name" in one of section in your project template')
        if tasks:
            [self._validate_task(task) for task in tasks]

    def _validate_project(self):
        name = self.project.get('name', None)
        color = self.project.get('color', None)
        tasks = self.project.get('tasks', [])
        sections = self.project.get('sections', [])

        if not name:
            raise ValidationError('There is no "name" in your project template')
        if color and int(color) not in range(30, 50):
            raise ValidationError('Project with name: "{}" should have color between: 30-49'.format(name))
        if not tasks and not sections:
            raise ValidationError('There is no "tasks" or "sections" in your "{}" project template'.format(name))
        if tasks:
            [self._validate_task(task) for task in tasks]
        if self.project.get('sections'):
            [self._validate_section(section) for section in sections]

    def validate(self):
        self._validate_project()


def validate_project(project_template):
    Validator(project_template).validate()
